lumen/ai/agents.py

```diff
@@ -656,131 +598,70 @@ async def _create_valid_sql(
+            if SOURCE_TABLE_SEPARATOR in source_table:
+                if match := re.match(pattern, source_table):
+                    a_source_name, a_table = match.groups()
+                    a_source_obj = sources.get(a_source_name)
                     a_table = f"{SOURCE_TABLE_SEPARATOR}{a_source_name}{SOURCE_TABLE_SEPARATOR}{a_table}"
             else:
-                a_source = next(iter(sources))
+                a_source_obj = next(iter(sources.values()))
                 a_table = source_table
+            tables_to_source[a_table] = a_source_obj
+        return tables_to_source, chain_of_thought
 
-            tables_to_source[a_table] = a_source
-        return tables_to_source
-
-    def _update_spec(self, memory: _Memory, event: param.parameterized.Event):
-        memory['sql'] = event.new
+    def _drop_source_table_separator(self, content: str) -> str:
+        return re.sub(rf".*?{SOURCE_TABLE_SEPARATOR}", "", content)
 
     async def respond(
         self,
         messages: list[Message],
         render_output: bool = False,
         step_title: str | None = None,
     ) -> Any:
-        """
-        Steps:
-        1. Retrieve the current source and table from memory.
-        2. If the source lacks a `get_sql_expr` method, return `None`.
-        3. Fetch the schema for the current table using `get_schema` without min/max values.
-        4. Determine if a join is required by calling `_check_requires_joins`.
-        5. If required, find additional tables via `find_join_tables`; otherwise, use the current source and table.
-        6. For each source and table, get the schema and SQL expression, storing them in `table_schemas`.
-        7. Render the SQL prompt using the table schemas, dialect, join status, and table.
-        8. If a join is required, remove source/table prefixes from the last message.
-        9. Construct the SQL query with `_create_valid_sql`.
-        """
         sources = self._memory["sources"]
-        tables_to_source, tables_schema_str = await gather_table_sources(sources)
-        table, source, join_required = await self._select_relevant_table(messages, sources, tables_to_source, tables_schema_str)
-        if not hasattr(source, "get_sql_expr"):
-            return None
-
-        schema = await get_schema(source, table, include_min_max=True)
-        tables_to_source = {table: source}
-        if join_required is None:
-            join_required = await self._check_requires_joins(messages, table, schema, tables_schema_str)
-            if join_required is None:
-                # Bail if query was cancelled or errored out
-                return None
-            if join_required:
-                tables_to_source = await self.find_join_tables(messages)
+        tables_to_source, tables_schema_str = await gather_table_sources(sources, include_sep=True)
+        tables_to_source, comments = await self._find_tables(messages, tables_schema_str)
```

Comment:
I'm finding this can return no tables at all and then errors below.

Reply:
Maybe we can wrap `retry_llm_output`?